stats/opencensus/opencensus.go

```diff
@@ -0,0 +1,126 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package opencensus implements opencensus instrumentation code for gRPC-Go
+// clients and servers.
+package opencensus
+
+import (
+	"context"
+
+	"go.opencensus.io/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/internal"
+	"google.golang.org/grpc/stats"
+)
+
+var (
+	joinDialOptions = internal.JoinDialOptions.(func(...grpc.DialOption) grpc.DialOption)
+)
+
+// TraceOptions are the tracing options for opencensus instrumentation.
+type TraceOptions struct {
+	// Ts is the Sampler used for tracing.
+	Ts trace.Sampler
+	// DisableTrace determines whether traces are disabled for an OpenCensus
+	// Dial or Server option. will overwrite any global option setting.
+	DisableTrace bool
+}
+
+// DialOption returns a dial option which enables OpenCensus instrumentation
+// code for a grpc.ClientConn.
+//
+// Client Applications interested in instrumenting their grpc.ClientConn should
+// pass the dial option returned from this function as the first dial option to
+// grpc.Dial().
+//
+// Using this option will always lead to instrumentation, however in order to
+// use the data an exporter must be registered with the OpenCensus trace package
+// for traces and the OpenCensus view package for metrics. Client Side has
+// retries, so a Unary and Streaming Interceptor are registered to handle per
+// RPC traces/metrics, and a Stats Handler is registered to handle per RPC
+// attempt trace/metrics. These three components registered work together in
+// conjunction, and do not work standalone.
+func DialOption(to TraceOptions) grpc.DialOption {
+	return joinDialOptions(grpc.WithChainUnaryInterceptor(unaryInterceptor), grpc.WithChainStreamInterceptor(streamInterceptor), grpc.WithStatsHandler(&clientStatsHandler{to: to}))
+}
+
+// ServerOption returns a server option which enables OpenCensus instrumentation
+// code for a grpc.Server.
+//
+// Server applications interested in instrumenting their grpc.Server should
+// pass the server option returned from this function as the first argument to
+// grpc.NewServer().
+//
+// Using this option will always lead to instrumentation, however in order to
+// use the data an exporter must be registered with the OpenCensus trace package
+// for traces and the OpenCensus view package for metrics. Server Side does not
+// have retries, so a registered Stats Handler is the only component you need
+// for RPC traces/metrics.
```

Comment:
"... so a registered Stats Handler is the only option that is returned"?

Also it is probably important to call out that (at least for now) it's not possible to use this in conjunction with another stats handler dial/server option.

Reply:
Fair enough, will add, but this was never possible right? I added binary logger option in this PR: https://github.com/grpc/grpc-go/pull/5675, which added a second possible binary logger on top of the global at each binary log recording point.

Reply:
Correct, it was never possible.  But this is the first time we've ever provided a function that returns a `grpc.WithStatsHandler` `DialOption` (/ `ServerOption`), as opposed to the OC package which provides a `grpc/stats.Handler`, so it's important to mention that this sets a `stats.Handler` and no other one should be used on the same client/server (for now).  It seems like we will need to support multiple in the not-too-distant future for migration to OpenTelemetry, though.